tests/js/server/aql/aql-filter-projections.js

```diff
@@ -0,0 +1,282 @@
+        assertEqual(query[1], nodes[0].indexes[0].type, query);
+        assertEqual(query[2], nodes[0].filterProjections, query);
+        assertEqual(query[3], nodes[0].projections, query);
+        if (nodes[0].projections.length > 0) {
+          assertNotEqual(-1, plan.rules.indexOf(ruleName));
+        }
+       
+        // query must not contain any FILTER nodes
+        nodes = plan.nodes.filter(function(node) { return node.type === 'FilterNode'; });
+        assertEqual(0, nodes.length);
+      });
+    },
+    
+    testPersistentSubAttribute : function () {
+      // no filter projections will be used for these queries
+      c.ensureIndex({ type: "persistent", fields: ["foo.bar"] });
+      let queries = [
+        [`FOR doc IN ${cn} FILTER doc.foo.bar == 1 RETURN doc.foo.bar`, 'persistent', [['foo', 'bar']] ],
+        [`FOR doc IN ${cn} FILTER doc.foo.bar == 1 RETURN doc.value`, 'persistent', ['value'] ],
+        [`FOR doc IN ${cn} FILTER doc.foo.bar == 1 RETURN doc`, 'persistent', [] ],
+      ];
+
+      queries.forEach(function(query) {
+        let plan = AQL_EXPLAIN(query[0], null, { optimizer: { rules: ["-optimize-cluster-single-document-operations"] } }).plan;
+        let nodes = plan.nodes.filter(function(node) { return node.type === 'IndexNode'; });
+        assertEqual(1, nodes.length, query);
+        assertEqual(1, nodes[0].indexes.length, query);
+        assertEqual(query[1], nodes[0].indexes[0].type, query);
+        assertEqual(query[2], nodes[0].projections, query);
+        assertEqual([], nodes[0].filterProjections, query);
+        if (nodes[0].projections.length > 0) {
+          assertNotEqual(-1, plan.rules.indexOf(ruleName));
+        }
+      });
+    },
+    
+    testPersistentSubAttributesMulti : function () {
+      c.ensureIndex({ type: "persistent", fields: ["foo.bar", "foo.baz"] });
+      let queries = [
+        [`FOR doc IN ${cn} FILTER doc.foo.bar == 1 FILTER doc.foo.baz != 1 RETURN doc.value`, 'persistent', [['foo', 'baz']], ['value'] ],
+        [`FOR doc IN ${cn} FILTER doc.foo.bar == 1 FILTER doc.foo.baz != 1 RETURN [doc.foo.bar, doc.value]`, 'persistent', [['foo', 'baz']], [['foo', 'bar'], 'value'] ],
+      ];
+
+      queries.forEach(function(query) {
+        let plan = AQL_EXPLAIN(query[0], null, { optimizer: { rules: ["-optimize-cluster-single-document-operations"] } }).plan;
+        let nodes = plan.nodes.filter(function(node) { return node.type === 'IndexNode'; });
+        assertEqual(1, nodes.length, query);
+        assertEqual(1, nodes[0].indexes.length, query);
+        assertEqual(query[1], nodes[0].indexes[0].type, query);
+        assertEqual(query[2], nodes[0].filterProjections, query);
+        assertEqual(query[3], nodes[0].projections, query);
+        assertNotEqual(-1, plan.rules.indexOf(ruleName));
+      });
+    },
+    
+    testPersistentStoredValues : function () {
+      c.ensureIndex({ type: "persistent", fields: ["value1"], storedValues: ["value2"] });
```

Comment:
Maybe add another test with sub-attributes as storedValues or even a combination of sub-attributes as indexed and stored values.